src/dos/dosirq.c

```diff
@@ -0,0 +1,312 @@
+/* Implementation of DOS IRQ routines, from libMikMod.
+   Copyright (C) 1999 by Andrew Zabolotny <bit@eltech.ru>
+ *
+ * Extended Module Player
+ * Copyright (C) 1996-2021 Claudio Matsuoka and Hipolito Carraro Jr
+ *
+ * This file is part of the Extended Module Player and is distributed
+ * under the terms of the GNU General Public License. See the COPYING
+ * file for more information.
+ */
+
+#include "dosirq.h"
+
+#include <dpmi.h>
+#include <go32.h>
+#include <dos.h>
+#include <sys/nearptr.h>
+#include <malloc.h>
+#include <string.h>
+
+unsigned int __irq_stack_size = 0x4000;
+unsigned int __irq_stack_count = 1;
+
+static void __int_stub_template (void)
+{
+/* *INDENT-OFF* */
+	asm("	pushal\n"
+		"	pushl	%ds\n"
+		"	pushl	%es\n"
+		"	pushl	%fs\n"
+		"	pushl	%gs\n"
+		"	movw	$0x1234,%ax\n"		/* Get DPMI data selector */
+		"	movw	%ax,%ds\n"			/* Set DS and ES to data selector */
+		"	movw	%ax,%es\n"
+		"	movl	$0x12345678,%ebx\n"	/* Interrupt stack top */
+		"	movl	(%ebx),%ecx\n"
+		"	movl	%ecx,%edx\n"
+		"	subl	$0x12345678,%ecx\n"	/* Subtract irq_stack_count */
+		"	movl	%ecx,(%ebx)\n"
+		"	movw	%ss,%si\n"			/* Save old SS:ESP */
+		"	movl	%esp,%edi\n"
+		"	movl	%edx,%esp\n"		/* Set SS:ESP to interrupt stack */
+		"	movw	%ax,%ss\n"
+		"	pushl	%esi\n"
+		"	pushl	%edi\n"
+		"	pushl	%ebx\n"
+		"	pushl	%edx\n"
+		"	call	1f\n"				/* Call user interrupt handler */
+		"1:	popl	%edx\n"
+		"	popl	%ebx\n"
+		"	movl	%edx,(%ebx)\n"
+		"	popl	%edi\n"
+		"	popl	%esi\n"
+		"	movl	%edi,%esp\n"		/* Restore old SS:ESP */
+		"	movw	%si,%ss\n"
+		"	popl	%gs\n"
+		"	popl	%fs\n"
+		"	popl	%es\n"
+		"	popl	%ds\n"
+		"	popal\n"
+		"	iret\n");
+/* *INDENT-ON* */
+}
+
+#include <stdio.h>
+
+static int _allocate_iret_wrapper(_go32_dpmi_seginfo * info)
```

Comment:
Is there a reason for using custom versions of `_allocate_iret_wrapper` and `_free_iret_wrapper`? The code seems to work fine with `_go32_dpmi_allocate_iret_wrapper` and `_go32_dpmi_free_iret_wrapper` from the standard library.

Reply:
Sorry for replying late.

As for the question: honestly I don't know - I'm not the original
author of the code, I only fixed the SB16 playback in there. If you
are absolutely confident about it I can change it.